sklearn/tree/_tree.pyx

```diff
@@ -107,19 +107,7 @@ cdef class TreeBuilder:
             # since we have to copy we will make it fortran for efficiency
             X = np.asfortranarray(X, dtype=DTYPE)
 
-        # TODO: This check for y seems to be redundant, as it is also
-        #  present in the BaseDecisionTree's fit method, and therefore
-        #  can be removed.
-        if y.base.dtype != DOUBLE or not y.base.flags.contiguous:
-            y = np.ascontiguousarray(y, dtype=DOUBLE)
-
-        if (
-            sample_weight is not None and
-            (
-                sample_weight.base.dtype != DOUBLE or
-                not sample_weight.base.flags.contiguous
-            )
-        ):
+        if sample_weight is not None and not sample_weight.base.flags.contiguous:
```

Comment:
I guess this `if` and the next line can be removed as well, right?

`sample_weight` is 1d so contiguous.

Reply:
Thinking about it and playing with `%%cython`, not necessarily since its definition is `const float64_t[:] sample_weight` and not `const float64_t[::1] sample_weight`

Reply:
> sample_weight is 1d so contiguous.

It can be 1D but not contiguous :)
```
x = np.arange(6)
y = x[::2]
y.flags
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
```

Reply:
> It can be 1D but not contiguous :)

This fact about 1d memoryviews come up from time to time. Feels like we need to write it down in https://scikit-learn.org/dev/developers/cython.html.